index/scorch/persister.go

```diff
@@ -266,16 +266,20 @@ func (s *Scorch) persistSnapshot(snapshot *IndexSnapshot) error {
 		for k, v := range s.root.internal {
 			newIndexSnapshot.internal[k] = v
 		}
-		for _, filename := range filenames {
-			delete(s.ineligibleForRemoval, filename)
-		}
+
 		rootPrev := s.root
 		s.root = newIndexSnapshot
 		s.rootLock.Unlock()
 		if rootPrev != nil {
 			_ = rootPrev.DecRef()
 		}
 	}
+	// unlock the files for clean up
+	s.rootLock.Lock()
+	for _, filename := range filenames {
+		delete(s.ineligibleForRemoval, filename)
+	}
+	s.rootLock.Unlock()
```

Comment:
I'm wondering about the case mentioned in the comment in line 222... what if only the internal keys were updated and needed to be persisted (line 266)?

wondering instead if the if-check on line 223 should just be removed?

Reply:
yea.. But removing the 223 if-check looked risky as I read the commit msg from @mschoch for it. Looked like this is the safest / easiest option.

Reply:
@steveyen your concern is valid, but perhaps we open separate issue to track that, as it appears to be a logical flaw with the way this works.   Specifically, the loop-breaking condition this relies on is incompatible with the case you point out.

Reply:
@mschoch - thx -- I'm good with the idea to separate that other feedback into a different issue / track.  (To be clear, that's the issue of "what if only the internal keys were updated and needed to be persisted" and whether we're handling that internal-keys-only correctly.)

Reply:
So I did break this out as a separate issue #763.  However I now believe it is correct, and the comment was trying to explain why.  The persister DID persist the snapshot containing the internal values.  What it is NOT doing is introducing a new root.  And the reason is that it does not have to.  The only reason to introduce a new root is if we want to swap in some newly persisted zap segments, but we don't have any of those in this case.  The new internal values are already available to readers through the work done by the introducer, and the persister did it's job to persist them in the root bolt with this snapshot.  So, for the case of internal only changes, there is nothing more to be done.